clippy_lints/src/casts/raw_slice_pointer_cast.rs

```diff
@@ -0,0 +1,62 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::{snippet, snippet_with_applicability};
+use clippy_utils::{match_def_path, meets_msrv, msrvs, paths};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{def_id::DefId, Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty::{self, Ty};
+use rustc_semver::RustcVersion;
+
+use super::RAW_SLICE_POINTER_CAST;
+
+enum RawPartsKind {
+    Immutable,
+    Mutable,
+}
+
+fn raw_parts_kind(cx: &LateContext<'_>, did: DefId) -> Option<RawPartsKind> {
+    if match_def_path(cx, did, &paths::SLICE_FROM_RAW_PARTS) {
+        Some(RawPartsKind::Immutable)
+    } else if match_def_path(cx, did, &paths::SLICE_FROM_RAW_PARTS_MUT) {
+        Some(RawPartsKind::Mutable)
+    } else {
+        None
+    }
+}
+
+pub(super) fn check(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    cast_expr: &Expr<'_>,
+    cast_to: Ty<'_>,
+    msrv: Option<RustcVersion>,
+) {
+    if_chain! {
+        if meets_msrv(msrv, msrvs::PTR_SLICE_RAW_PARTS);
+        if let ty::RawPtr(ptrty) = cast_to.kind();
+        if let ty::Slice(_) = ptrty.ty.kind();
+        if let ExprKind::Call(fun, [ptr_arg, len_arg]) = cast_expr.peel_blocks().kind;
+        if let ExprKind::Path(ref qpath) = fun.kind;
+        if let Some(fun_def_id) = cx.qpath_res(qpath, fun.hir_id).opt_def_id();
+        if let Some(rpk) = raw_parts_kind(cx, fun_def_id);
+        then {
+            let func = match rpk {
+                RawPartsKind::Immutable => "from_raw_parts",
+                RawPartsKind::Mutable => "from_raw_parts_mut"
+            };
+            let span = expr.span;
+            let mut applicability = Applicability::MachineApplicable;
+            let func_snippet = snippet_with_applicability(cx, fun.span, func, &mut applicability);
+            let ptr = snippet(cx, ptr_arg.span, "ptr");
+            let len = snippet(cx, len_arg.span, "len");
+            span_lint_and_sugg(cx,
+                RAW_SLICE_POINTER_CAST,
+                span,
+                &format!("casting the result of `{}` to {cast_to}", func_snippet),
+                "replace with",
+                format!("core::ptr::slice_{func}({ptr}, {len})"),
+                applicability);
```

Comment:
help/suggestion message + formatting nit:

```suggestion
            let ptr = snippet_with_applicability(cx, ptr_arg.span, "ptr", &mut applicability);
            let len = snippet_with_applicability(cx, len_arg.span, "len", &mut applicability);
            span_lint_and_sugg(
                cx,
                RAW_SLICE_POINTER_CAST,
                span,
                &format!("casting the result of `{func}` to {cast_to}", func_snippet),
                "replace with",
                format!("core::ptr::slice_{func}({ptr}, {len})"),
                applicability
            );
```